File: clients/apple/MultiLiveTV/Services/SourceHealthStore.swift
import Foundation

struct SourceRuntimeStatus: Hashable {
    var healthy: Bool
    var lastSuccess: Date?
    var errorCount: Int

    static let initial = SourceRuntimeStatus(healthy: true, lastSuccess: nil, errorCount: 0)
}

/// Process-local health for collectable sources. Never mutates `Source.enabled`.
final class SourceHealthStore: @unchecked Sendable {
    static let shared = SourceHealthStore()
    static let failureThreshold = 5
    /// Applied to play ranking when unhealthy (negative).
    static let unhealthyPenalty = 200

    private let lock = NSLock()
    private var statusByNumericId: [Int: SourceRuntimeStatus] = [:]

    init() {}

    func status(for numericId: Int) -> SourceRuntimeStatus {
        lock.lock()
        defer { lock.unlock() }
        return statusByNumericId[numericId] ?? .initial
    }

    func recordSuccess(numericId: Int, at date: Date = Date()) {
        lock.lock()
        defer { lock.unlock() }
        statusByNumericId[numericId] = SourceRuntimeStatus(
            healthy: true,
            lastSuccess: date,
            errorCount: 0
        )
    }

    func recordFailure(numericId: Int) {
        lock.lock()
        defer { lock.unlock() }
        var current = statusByNumericId[numericId] ?? .initial
        current.errorCount += 1
        if current.errorCount >= Self.failureThreshold {
            current.healthy = false
        }
        statusByNumericId[numericId] = current
    }

    func healthScore(for numericId: Int) -> Int {
        let s = status(for: numericId)
        if !s.healthy { return -Self.unhealthyPenalty }
        if s.errorCount == 0 { return 0 }
        return -min(s.errorCount * 20, Self.unhealthyPenalty - 20)
    }

    func reset() {
        lock.lock()
        defer { lock.unlock() }
        statusByNumericId.removeAll()
    }
}

File: clients/android/core/src/main/kotlin/com/heibaimiao/multilivetv/source/SourceHealthStore.kt
package com.heibaimiao.multilivetv.source

data class SourceRuntimeStatus(
    val healthy: Boolean = true,
    val lastSuccessMillis: Long? = null,
    val errorCount: Int = 0,
) {
    companion object {
        val initial = SourceRuntimeStatus()
    }
}

class SourceHealthStore {
    private val lock = Any()
    private val statusByNumericId = mutableMapOf<Int, SourceRuntimeStatus>()

    fun status(numericId: Int): SourceRuntimeStatus = synchronized(lock) {
        statusByNumericId[numericId] ?: SourceRuntimeStatus.initial
    }

    fun recordSuccess(numericId: Int, atMillis: Long = System.currentTimeMillis()) {
        synchronized(lock) {
            statusByNumericId[numericId] = SourceRuntimeStatus(
                healthy = true,
                lastSuccessMillis = atMillis,
                errorCount = 0,
            )
        }
    }

    fun recordFailure(numericId: Int) {
        synchronized(lock) {
            val current = statusByNumericId[numericId] ?: SourceRuntimeStatus.initial
            val errorCount = current.errorCount + 1
            statusByNumericId[numericId] = current.copy(
                errorCount = errorCount,
                healthy = errorCount < FAILURE_THRESHOLD,
            )
        }
    }

    fun healthScore(numericId: Int): Int {
        val s = status(numericId)
        if (!s.healthy) return -UNHEALTHY_PENALTY
        if (s.errorCount == 0) return 0
        return -minOf(s.errorCount * 20, UNHEALTHY_PENALTY - 20)
    }

    fun reset() {
        synchronized(lock) { statusByNumericId.clear() }
    }

    companion object {
        const val FAILURE_THRESHOLD = 5
        const val UNHEALTHY_PENALTY = 200
        val shared = SourceHealthStore()
    }
}
